### PR TITLE
[SPARK-30969][CORE] Remove resource coordination support from Standalone

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Remove automatically resource coordination support from Standalone.

### Why are the changes needed?

Resource coordination is mainly designed for the scenario where multiple workers launched on the same host. However, it's, actually, a non-existed  scenario for today's Spark. Because, Spark now can start multiple executors in a single Worker, while it only allow one executor per Worker at very beginning. So, now, it really help nothing for user to launch multiple workers on the same host. Thus, it's not worth for us to bring over complicated implementation and potential high maintain cost for such an impossible scenario.

### Does this PR introduce any user-facing change?

No, it's Spark 3.0 feature.

### How was this patch tested?

Pass Jenkins.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @mengxr  @tgravescs @jiangxb1987 @cloud-fan

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by tgravescs** [comment]
Can you please clarify the overall behavior changes? Its just running discovery script and assuming this owns all of them, correct?
I believe this will break local-cluster mode unless you have handled in your change? local cluster mode you can’t specify different scripts/resource files per worker so how is this now handled?
This also handled coordination between multiple drivers (if you gave drivers gpus) and workers on the same node.
I know of people running multiple workers per node - so please provide more justification for this - is it causing performance issue? It seems like this provided useful feature for so unless its causing some issue why remove.

**Comment by Ngone51** [comment]
> Its just running discovery script and assuming this owns all of them, correct?

Yea, I think this is the only difference. And it really break local-cluster mode. Though, I haven't made change in `LocalClusterSpark` to handle it and only update one related test to workaround it. And there's no more local-cluster mode tests rely on ***coordination**.

And, of cause, user should still be aware of resource conflict between multi drivers. So I still keep the warning in the doc. Same node driver and worker mode, actually, is not recommended when driver submitted in client mode. Ideally, it should be an edge node out of cluster for the driver and assumed not to be on the same of host of worker.

**Comment by tgravescs** [comment]
>  Ideally, it should be an edge node out of cluster for the driver and assumed not to be on the same of host of worker.
This may be ideal but not true in many cases, especially in just test/dev setups.

For me this comes down to making things convenient for the user and doesn't seem like much overhead on the Spark side, especially since its already implemented/tested.  If you aren't using custom resources, the code has no affect (or shouldn't anyway - if it does we should fix). Its configurable off if people don't want it or see some overhead from the file locking pieces. The code for this in my opinion is not that much and pretty isolated. I don't see much downside here.

I guess the spark.resources.dir could be left around in worker crash type cases. Though again could be turned off. 

The only way I see to fix the local-cluster mode is leave this code or somehow allow users to specify different resource files or script per worker, which seems like a hard/weird api. 

I would like more justification here? why is this a high maintenance cost? the code is done and tested and I don't envision it needed more feature work. local-cluster mode unit tests should be enough to test it pretty thoroughly (perhaps we need more).

**Comment by mengxr** [comment]
@tgravescs I suggested removal of this resource coordination code in an offline 3.0 API audit discussion to keep Spark codebase simple. Scheduling is already an area that lacks of maintainers. Increasing its complexity would keep potential maintainers away. This is why we have been very careful when we introduced the resource-aware scheduling feature. Here are the reasons to remove this feature before Spark 3.0 release:

* In a real standalone deployment, having multiple workers on the same host is not longer needed. As @Ngone51 mentioned, the feature was introduced to keep worker JVM small to avoid long GC pause when worker and executors share the same JVM. Now executors run in a different JVM, so having one worker per host is sufficient. Correct me if I'm wrong since you mentioned several deployment using multiple workers on the same host.
* Due to the reason above, instead of supporting GPU scheduling for multiple workers on the same host, we should deprecate entirely the support of multiple workers on the same host in 3.0 and remove it in a future release.
* The local-cluster mode is not a public feature, which should only be used in Spark tests. Actually this change is the first to mention "local-cluster" in the user guide, which makes it "public". I don't think we want to add (even localized) complexity just for this mode. In a test setup, we can use one worker process and separate driver/worker scripts to simplify the resource allocation. Again, using multiple workers in local-cluster mode is to simulate a real cluster, not a "production" setup. If this is for simulation, we just need fake discovery scripts.
* For worker recovery, my understanding is that there shouldn't be a case that the old worker and the recovery worker are running at the same time. Because the recovery is usually done by monit which monitors the worker process. cc: @jiangxb1987 

cc: @squito

**Comment by cloud-fan** [comment]
I'm not very familiar with the whole story, but the recent discussion of backward compatibility makes me worry about this one.

If we release a very complicated API for 1% of the user cases, it may not worth the cost to maintain that API. Already implemented/tested doesn't count the future maintenance cost. Recently I'm suffered from keeping the support of some legacy behavior/API and I hope we can avoid such problems at the beginning.

Are there any valid use cases for multi-worker per host? If not or only a little, I'd suggest we at least don't support it in the new GPU scheduling.

**Comment by jiangxb1987** [comment]
> I know of people running multiple workers per node

@tgravescs Could you please share details about those workloads? Based on our experience it's not necessarily needed to deploy multiple workers on the same host.

**Comment by tgravescs** [comment]
Ok, after thinking about this some more, I'm ok with removing. its a new feature and if we are really pushing to support only one worker per node it makes sense to not add it in.  I think you deal with standalone mode in production more then me and most of the cases I'm thinking of are dev/testing type cases. It can be worked around by specifying different resource files for different workers (except for local-cluster) so shouldn't be an issue with a little extra setup if people do want to run multiple workers per node.
We need to explicitly state that local-cluster mode is not supported with custom resources.

File: docs/configuration.md [hunk 1]
```diff
@@ -2782,5 +2762,4 @@ There are configurations available to request resources for the driver: <code>sp
...
 
-See your cluster manager specific page for requirements and details on each of - [YARN](running-on-yarn.html#resource-allocation-and-configuration-overview), [Kubernetes](running-on-kubernetes.html#resource-allocation-and-configuration-overview) and [Standalone Mode](spark-standalone.html#resource-allocation-and-configuration-overview). It is currently not available with Mesos or local mode. If using local-cluster mode see the Spark Standalone documentation but be aware only a single worker resources file or discovery script can be specified the is shared by all the Workers so you should enable resource coordination (see <code>spark.resources.coordinateResourcesInStandalone</code>).
-
+See your cluster manager specific page for requirements and details on each of - [YARN](running-on-yarn.html#resource-allocation-and-configuration-overview), [Kubernetes](running-on-kubernetes.html#resource-allocation-and-configuration-overview) and [Standalone Mode](spark-standalone.html#resource-allocation-and-configuration-overview). It is currently not available with Mesos or local mode. If using local-cluster mode please see the Spark Standalone documentation.
```
Comment (tgravescs) [inline docs/configuration.md]: we should modify to say local-cluster mode not supported - I guess if we want to be very specific we could state say with multiple workers (but I'm fine just saying not supported in general).

**Comment by tgravescs** [comment]
test this please

File: core/src/main/scala/org/apache/spark/deploy/StandaloneResourceUtils.scala [hunk 2]
```diff
@@ -17,30 +17,21 @@
...
 
 import scala.collection.mutable
-import scala.util.Random
 import scala.util.control.NonFatal
 
 import org.json4s.{DefaultFormats, Extraction}
-import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.jackson.JsonMethods.{compact, render}
 
 import org.apache.spark.{SparkConf, SparkException}
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/StandaloneResourceUtils.scala]: SparkConf no longer needed

**Review by tgravescs — COMMENTED** [review]
one small nit, otherwise looks good. I think it would be good to run a few manual tests with the resourcefile and discovery script just to make sure we didn't miss anything

**Comment by Ngone51** [comment]
@tgravescs thanks for the review. I manually tested with resources file and script locally and checked in UI. It looks good.

**Review by tgravescs — APPROVED** [review]

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
Thanks, merged to master/3.0 !

**Comment by Ngone51** [comment]
Thanks all!!

**Pull request closed without merging**